server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -160,4 +384,478 @@ private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request)
+                        t -> new NodeQueryRequest(request, numberOfShardsTotal, timeProvider.absoluteStartMillis(), t.clusterAlias())
+                    );
+                    final String indexUUID = routing.getShardId().getIndex().getUUID();
+                    perNodeRequest.shards.add(
+                        new ShardToQuery(
+                            concreteIndexBoosts.getOrDefault(indexUUID, DEFAULT_INDEX_BOOST),
+                            getOriginalIndices(shardIndex),
+                            shardIndex,
+                            routing.getShardId(),
+                            shardRoutings.getSearchContextId()
+                        )
+                    );
+                    var filterForAlias = aliasFilter.getOrDefault(indexUUID, AliasFilter.EMPTY);
+                    if (filterForAlias != AliasFilter.EMPTY) {
+                        perNodeRequest.aliasFilters.putIfAbsent(indexUUID, filterForAlias);
+                    }
+                } else {
+                    performPhaseOnShard(shardIndex, shardRoutings, routing);
+                }
+            }
+        }
+        perNodeQueries.forEach((routing, request) -> {
+            if (request.shards.size() == 1) {
+                executeAsSingleRequest(routing, request.shards.getFirst());
+                return;
+            }
+            final Transport.Connection connection;
+            try {
+                connection = getConnection(routing.clusterAlias(), routing.nodeId());
+            } catch (Exception e) {
+                onNodeQueryFailure(e, request, routing);
+                return;
+            }
+            // must check both node and transport versions to correctly deal with BwC on proxy connections
+            if (connection.getTransportVersion().before(TransportVersions.BATCHED_QUERY_PHASE_VERSION)
+                || connection.getNode().getVersionInformation().nodeVersion().before(Version.V_9_1_0)) {
+                executeWithoutBatching(routing, request);
+                return;
+            }
+            searchTransportService.transportService()
+                .sendChildRequest(connection, NODE_SEARCH_ACTION_NAME, request, task, new TransportResponseHandler<NodeQueryResponse>() {
+                    @Override
+                    public NodeQueryResponse read(StreamInput in) throws IOException {
+                        return new NodeQueryResponse(in);
+                    }
+
+                    @Override
+                    public Executor executor() {
+                        return EsExecutors.DIRECT_EXECUTOR_SERVICE;
+                    }
+
+                    @Override
+                    public void handleResponse(NodeQueryResponse response) {
+                        if (results instanceof QueryPhaseResultConsumer queryPhaseResultConsumer) {
+                            queryPhaseResultConsumer.addPartialResult(response.topDocsStats, response.mergeResult);
```

Comment:
the instanceof checks requires in this class around query phase result consumer make me wonder if something is off. When are results not falling into this category in practice and what happens then? Are these scenario covered in tests?

Reply:
The only other thing you could see here is the `org.elasticsearch.action.search.CountOnlyQueryPhaseResultConsumer`. I had that one have batched execution logic before as well but it increases the size of the PR somewhat.
Since the partial reduction logic on the data node side doesn't touch the `totalHits`, this is all that's needed to make the other consumer run fine transparently.
I think as a follow-up we could obviously just make it support batched execution more explicity (and even more performant, there's no reason whatsoever it even needs any of the query-results, it only needs count + failed shards back from the data node). For now this was just the shortest possible change :)